hnn_qt5.py

```diff
@@ -218,71 +231,174 @@ def runsim (self):
+    os.makedirs(optinf_dir, exist_ok=True)
+    fpopt = open(fnoptinf,'w')
     fpopt.close()
     self.minopterr = 1e9
+    self.stepminopterr = self.minopterr
+    self.best_ddat = None
 
-    def optrun (vparam):
-      # create parameter dictionary of current values to test
-      lparam = list(dconf['params'].values())
+    def optrun (new_params, grad=0):
       dtest = OrderedDict() # parameter values to test      
-      for prm,val in zip(lparam,expvals(vparam,lparam)): # set parameters
-        if val >= prm.minval and val <= prm.maxval:
-          if debug: print('optrun prm:',prm.var,prm.minval,prm.maxval,val)
-          dtest[prm.var] = val
+      for param_name, test_value in zip(self.opt_params['ranges'].keys(), new_params): # set parameters
+        if test_value >= self.opt_params['ranges'][param_name]['minval'] and test_value <= self.opt_params['ranges'][param_name]['maxval']:
+          if debug:
+            print('optrun prm:', self.opt_params['ranges'][param_name]['initial'],
+                                 self.opt_params['ranges'][param_name]['minval'],
+                                 self.opt_params['ranges'][param_name]['maxval'],
+                                 test_value)
+          dtest[param_name] = test_value
         else:
-          if debug: print('optrun:', val, 'out of bounds for ' , prm.var, prm.minval, prm.maxval)
+          if debug:
+            print('optrun:', test_value, 'out of bounds for ' ,
+                  self.opt_params['ranges'][param_name]['initial'],
+                  self.opt_params['ranges'][param_name]['minval'],
+                  self.opt_params['ranges'][param_name]['maxval'])
           return 1e9 # invalid param value -> large error
-      if debug:
-        if type(vparam)==list: print('set params:', vparam)
-        else: print('set params:', vparam.as_numpy())
+
+      # stop the sim early if possible
+      dtest['tstop'] = self.opt_params['opt_end']
 
       self.updatebaseparamwin(dtest) # put new param values into GUI
       sleep(1)
 
-      self.runsim() # run the simulation as usual and read its output
-      import simdat
-      simdat.calcerr(simdat.ddat) # make sure error re-calculated (synchronously)
-      self.updatedrawerr() # send event to draw updated error (asynchronously)
+      self.runsim(is_opt=True) # run the simulation as usual and read its output
+
+      if self.last_step:
+        simdat.weighted_rmse(simdat.ddat,
+                            self.opt_params['opt_end'],
+                            self.opt_params['weights'],
+                            tstart=self.opt_params['opt_start'])
+        simdat.calcerr(simdat.ddat,
+                      self.opt_params['opt_end'],
+                      tstart=self.opt_params['opt_start'])
+        err = simdat.ddat['werrtot']
+#        print("regular RMSE = %f"% (simdat.ddat['errtot']))
+        print("weighted RMSE = %f, regular RMSE = %f"% (simdat.ddat['werrtot'],simdat.ddat['errtot']))
```

Comment:
TODO: this is a mistake. Should not print weighted RMSE for last step. I verified that regular RMSE is the same as wRMSE with weights of all 1's.